internal/guest/runtime/hcsv2/container.go

```diff
@@ -181,3 +185,20 @@ func (c *Container) GetStats(ctx context.Context) (*v1.Metrics, error) {
 func (c *Container) modifyContainerConstraints(ctx context.Context, rt prot.ModifyRequestType, cc *prot.ContainerConstraintsV2) (err error) {
 	return c.Update(ctx, cc.Linux)
 }
+
+// adjustExecSpec adjusts an OCI runtime specs process field provided for an execed process to contain some of the settings set on the containers spec.
+// Some of the OCI spec settings are configured in the guest as there's not enough information available on the host to set them, so the spec set for the
+// exec may be missing some things that were present in the final version of the containers spec.
+func adjustExecSpec(containerSpec *oci.Process, execSpec *oci.Process) {
+	// One of the aforementioned settings is the containers user configured for the init process as we can't verify that the user exists in
+	// the container image on the host. A consequence of the user example is any exec will run as root instead of the user configured for
+	// the container which can be problematic.
+	//
+	// There is no way to tell if a client supplied root as the user for an exec explicitly or if they simply didn't fill in
+	// the spec, as the default values of uid 0 and gid 0 are also the uid & gid values of the root user. To try and cover as many
+	// use cases as possible, we check if the uid/gid are changed at all and if not we just copy whatever user was set in the
+	// container spec to the exec spec.
+	if execSpec.User.UID == 0 && execSpec.User.GID == 0 {
```

Comment:
This seems like a dangerous assumption to make if gcs is ever used outside of our cri fork.

Reply:
True, I wish user was a pointer field so we'd have a better case to go off of. Judging from your above comment and me revisiting the code, it might be possible to set it on the spec earlier and skip the rootfs checking bits for lcow.